src/logic/algo/solver.ts

```diff
@@ -23,20 +23,124 @@ export type StepToImplementName = keyof Solver;
 export class MonteCarloTreeSearch implements Solver {
   private kernel: NonNullable<MonteCarloTreeSearchHyperParams["kernel"]> = ubcKernel;
 
+  private nodeRewards: Map<TreeNode, number> = new Map();
+
+  private nodeVisits: Map<TreeNode, number> = new Map();
+
+  private nodeChildren: Map<TreeNode, Array<TreeNode>> = new Map();
+
   run(root: TreeNode, hyperParams: MonteCarloTreeSearchHyperParams): TreeNode {
+    while (true) {
+      for (let i = 0; i < hyperParams.numIterations; i += 1) {
+        this.singleRun(root, hyperParams);
+      }
+      root = this.choose(root);
+      if (root.children.length === 0) {
+        return root;
+      }
+    }
+  }
+
+  singleRun(root: TreeNode, hyperParams: MonteCarloTreeSearchHyperParams): void {
     this.kernel = hyperParams.kernel ?? this.kernel;
-    return new TreeNode();
+    const path = this.select(root);
+    const leaf = path[path.length - 1];
+    this.expand(leaf);
+    let reward = 0;
+    for (let i = 0; i < hyperParams.numRollout; i += 1) {
+      reward += this.simulate(leaf);
+    }
+    this.backup(path, reward);
   }
 
-  choose(node: TreeNode): null {
-    return null;
+  select(node: TreeNode): Array<TreeNode> {
+    const path: Array<TreeNode> = [];
+    while (true) {
+      path.push(node);
+      if (!this.nodeChildren.has(node)) {
+        return path;
+      }
+      if (this.nodeChildren.get(node) ?? [].length === 0) {
+        return path;
+      }
+      let unexploredNodes: Array<TreeNode> = this.nodeChildren.get(node) ?? [];
+      unexploredNodes = unexploredNodes.filter(
+        (unexploredNode) => !this.nodeChildren.has(unexploredNode)
+      );
+      if (unexploredNodes.length > 0) {
+        const n: TreeNode | undefined = unexploredNodes.pop();
+        if (n) {
+          path.push(n);
+        }
+        return path;
+      }
+      node = this.uctSelect(node);
+    }
   }
 
-  select(node: TreeNode): null {
-    return null;
+  choose(node: TreeNode): TreeNode {
+    if (node.children.length === 0) {
+      throw Error("Choose method called on terminal node!");
```

Comment:
Czy takie rzucanie błędów jest "dozwolone"?

Reply:
Tak
Jeśli chodzi o składnię to można zarówno `new Error` jak i `Error`
Jeśli chodzi o logikę - to też można, bo to jest coś czego nieoczekujemy więc nie musimy tego obsługiwać